### ISSUE TITLE
[Feature] Startup Taints

### ISSUE THREAD

**Issue by allyford** [Azure/AKS#3464]
For node initialization scenarios, customers can use taints passed to kubelet on startup. These taints will not be reconciled by AKS at runtime.

**Comment by htpawel** [comment]
Any update on this? @allyford

**Comment by hterik** [comment]
related: https://github.com/Azure/AKS/issues/3276 https://github.com/Azure/AKS/issues/2934

**Comment by hterik** [comment]
Thanks for the good news @allyford. :+1:  
How does this feature interact with the cluster-autoscaler? 
[The documentation](https://learn.microsoft.com/en-gb/azure/aks/use-node-taints#node-taint-options) says it is useful for doing node-initialization.

> _Node initialization taints are placed on the node at boot time and are meant to be used temporarily, such as in **scenarios where you might need extra time to set up your nodes**._

I believe for that to work, the `startup-taints` option in the cluster-autoscaler also need to be configurable, see https://github.com/Azure/AKS/issues/3276. Otherwise the autoscaler might keep producing new nodes, when it sees that pods don't get scheduled, if the first new nodes take long time to set up and clear their init-taints.

Completing that ticket and some documentation on how to achieve such a scenario together with the autoscaler would be very useful.

**Comment by UtheMan** [comment]
@hterik - init taints added to the cluster spec are automatically added to the list of taints ignored by the AKS AutoScaler. We can add more details to docs to make it clearer.

**Comment by htpawel** [comment]
@UtheMan yes, please :) Is there any timeout for init taints btw?

**Comment by UtheMan** [comment]
@htpawel init taints are now available in preview [here](https://learn.microsoft.com/en-us/azure/aks/use-node-taints#use-node-initialization-taints-preview). We are working on removing some of the limitations that the preview version has and once that is out of the way we will begin the GA process for this feature.

**Comment by luistilingue** [comment]
It is an exciting preview feature, but it doesn't work well due to this issue related to system pool:

```
(SystemPoolHasRestrictedTaint) Placing custom taints on system pool is not supported(except 'CriticalAddonsOnly' taint or taint effect is 'PreferNoSchedule'). Please refer to https://aka.ms/aks/system-taints for detail
Code: SystemPoolHasRestrictedTaint
Message: Placing custom taints on system pool is not supported(except 'CriticalAddonsOnly' taint or taint effect is 'PreferNoSchedule'). Please refer to https://aka.ms/aks/system-taints for detail
Target: agentPoolProfile.nodeTaints
```

**Comment by allyford** [comment]
@luistilingue - Placing custom taints on the system node pool are still not supported. This feature does not enable system pool taints.

**Comment by luistilingue** [comment]
@allyford yes, but using the `az cli` I can't pass which nodepool I'm going to enable initialization taints, it enable in all cluster node pools.

Eg.

```
az aks update --resource-group acme --name xmtp --nodepool-initialization-taints "node.cilium.io/agent-not-ready=true:NoExecute"
```

**Comment by zhangchl007** [comment]
@luistilingue @allyford   Azure CLI  is not ready for this preview feature yet ? Please refer to the screen snapshot below. but I can update the nodeInitialization Taints via the cluster operation api

"nodeInitializationTaints": [
            "initTaintExample=value:NoSchedule"
        ]

**Comment by zhangchl007** [comment]
@hterik I passed the agent profile as below when I specify the node pool, Hope AZ-CLI could be fixed soon.
 "agent_pool_profiles": [
                {
                    "name": "nodepool1",
                    "count": 2,
                    "vm_size": "Standard_DS2_v2",
                    "max_pods": 250,
                    "os_type": "Linux",
                    "type": "VirtualMachineScaleSets",
                    "enable_auto_scaling": False,
                    "mode": "System",
                    "vnetSubnetID": "xxxxxx",
                    "podSubnetID": "xxxxxx"
                },
                {
                    "name": "testnodpool",
                    "count": 1,
                    "max_pods": 250,
                    "os_type": "Linux",
                    "osSKU": "AzureLinux",
                    "type": "VirtualMachineScaleSets",
                    "enable_auto_scaling": False,
                    "vm_size": "Standard_D4as_v5",
                    "mode": "User",
                    "vnetSubnetID": "xxxxxx",
                    "podSubnetID": "xxxxxx",
                    "nodeInitializationTaints": [
                      "sku=paas:NoSchedule"
                    ]
                }

$ az aks nodepool show --resource-group $RESOURCE_GROUP_NAME --cluster-name $CLUSTER_NAME --name $NODE_POOL_NAME --query "nodeInitializationTaints"
The behavior of this command has been altered by the following extension: aks-preview
[
  "sku=paas:NoSchedule"
]

NODE_POOL_NAME=spotazlinux1
:$ az aks nodepool show --resource-group $RESOURCE_GROUP_NAME --cluster-name $CLUSTER_NAME --name $NODE_POOL_NAME --query "nodeInitializationTaints"
The behavior of this command has been altered by the following extension: aks-preview

**Comment by zirkome** [comment]
Hi @allyford, We haven't heard from you this September, is there any update you can give for GA? Thank you!

**Comment by robbo10** [comment]
Is there any update on this?

The documentation is very misleading per - https://learn.microsoft.com/en-us/azure/aks/use-node-taints

```

The node initialization taints you specify apply to all of the node pools in the cluster. To apply the initialization taint to a specific node, you can use an ARM template instead of the CLI.

``` 

Yet throughout the comments above you are saying the system node pool only accepts the critical addons taint??

This is a very over complicated process for people looking to simply add/remove a taint on node start up.

**Comment by amitmavgupta** [comment]
Not stale.
